experimental/zerotrie/src/reader.rs

```diff
@@ -0,0 +1,454 @@
+        }
+    }
+}
+
+/// Given a slice starting with an offset table, returns the trie for the given index.
+///
+/// Arguments:
+/// - `trie` = a trie pointing at an offset table (after the branch node and search table)
+/// - `i` = the desired index within the offset table
+/// - `n` = the number of items in the offset table
+/// - `w` = the width of the offset table items minus one
+#[inline]
+fn get_branch(mut trie: &[u8], i: usize, n: usize, mut w: usize) -> Option<&[u8]> {
+    let mut p = 0usize;
+    let mut q = 0usize;
+    loop {
+        let indices;
+        (indices, trie) = debug_split_at(trie, n - 1)?;
+        p = (p << 8)
+            + if i == 0 {
+                0
+            } else {
+                debug_get(indices, i - 1)? as usize
+            };
+        q = match indices.get(i) {
+            Some(x) => (q << 8) + *x as usize,
+            None => trie.len(),
+        };
+        if w == 0 {
+            break;
+        }
+        w -= 1;
+    }
+    debug_get_range(trie, p..q)
+}
+
+/// Version of [`get_branch()`] specialized for the case `w == 0` for performance
+#[inline]
+fn get_branch_w0(mut trie: &[u8], i: usize, n: usize) -> Option<&[u8]> {
+    let indices;
+    (indices, trie) = debug_split_at(trie, n - 1)?;
+    let p = if i == 0 {
+        0
+    } else {
+        debug_get(indices, i - 1)? as usize
+    };
+    let q = match indices.get(i) {
+        Some(x) => *x as usize,
+        None => trie.len(),
+    };
+    debug_get_range(trie, p..q)
+}
+
+enum ByteType {
+    Ascii,
+    Span,
+    Value,
+    Match,
+}
+
+impl core::fmt::Debug for ByteType {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        use ByteType::*;
+        f.write_str(match *self {
+            Ascii => "a",
+            Span => "s",
+            Value => "v",
+            Match => "m",
+        })
+    }
+}
+
+#[inline]
+fn byte_type(b: u8) -> ByteType {
+    match b & 0b11100000 {
+        0b10000000 => ByteType::Value,
+        0b10100000 => ByteType::Span,
+        0b11000000 => ByteType::Match,
+        0b11100000 => ByteType::Match,
+        _ => ByteType::Ascii,
+    }
+}
+
+// DISCUSS: This function is 7% faster *on aarch64* if we assert a max on w.
+//
+// | Bench         | No Assert, x86_64 | No Assert, aarch64 | Assertion, x86_64 | Assertion, aarch64 |
+// |---------------|-------------------|--------------------|-------------------|--------------------|
+// | basic         | ~187.51 ns        | ~97.586 ns         | ~199.11 ns        | ~99.236 ns         |
+// | subtags_10pct | ~9.5557 µs        | ~4.8696 µs         | ~9.5779 µs        | ~4.5649 µs         |
+// | subtags_full  | ~137.75 µs        | ~76.016 µs         | ~142.02 µs        | ~70.254 µs         |
+
+/// Query the trie assuming all branch nodes are binary search.
+pub fn get_bsearch_only(mut trie: &[u8], mut ascii: &[u8]) -> Option<usize> {
+    loop {
+        let (b, x, i, search);
+        (b, trie) = trie.split_first()?;
```

Comment:
thought (nb): a useful thing for documentation may be to have a `RawTrie(&[u8])`  type with more descriptive methods on the operations being performed, e.g. this would be `.read_byte()` and the varint stuff could also live as methods. unsure of details

Reply:
Hmm. I don't really consider the reader function to be one in need of a refactor. In my earlier development, I wrote a few versions of the function and iterated until I got the code size and performance I wanted. I don't think this specific formulation came up in my earlier development, though.

Reply:
Mostly from a correctness standpoint: this code is somewhat stateful when it comes to whether the current cursor in the trie is itself a valid trie, and it seems like a useful distinction to maintain in code. But not blocking, like i said.